### PR TITLE
Prevent rename after single click

### PR DESCRIPTION
**Pull request by Don-Vito**
Closes: https://github.com/files-community/Files/issues/2231

Disclaimer: I really hate this solution, and myself.. for not finding something better :smile:

Root cause:
* When the user clicks outside of the AllView datagrid we manually clean the selection
* But from the datagrid perspective the "current cell" did not change
* Hence clicking on this "current cell" again triggers editing
(as per data grid specification)

I didn't find any reasonable way to convince the datagrid that the current
slot is invalid, hence managed this manually by "remembering" the click
outside of the grid. If such click was performed we consider there is
no current cell, unless the user actively clicks on one of the cells.

### CODE REVIEW

**Comment by gave92** [comment]
@Don-Vito thanks for this. With this PR all the stuff related to `GenericFileBrowser.xaml.cs -> tapDebounceTimer` can probably be remove

**Review by gave92 — COMMENTED** [review]
@Don-Vito thanks for this. With this PR all the stuff related to [GenericFileBrowser.xaml.cs -> tapDebounceTimer](https://github.com/files-community/Files/pull/2762/files#diff-84c054f6ded97591cf9ef1d64a6c0d5348723d507fe5b89d4d8e80e8d9d59a34R275) can probably be removed (it was another ugly solution for this issue)

**Comment by Don-Vito** [comment]
@gave92 - I would avoid nominating this PR to the beauty contest as well 😆 
I will check if debouncer can be removed.

Probably there is a better solution, or probably we need to extend the data grid / contribute to it.
Since this issue is just a "one of" example. There are more, e.g., if you start shift selection in the grid
and click very fast we identify this as rectangle selection (because there is both mouse move and the button is pressed).
Our rectangle selection resets the selected items, effectively canceling the shift selection of the grid.

This could be avoided if we would extend the grid and implement the rectangle selection and current cell deselection "natively".
WDYT?

File: Files/Views/LayoutModes/GenericFileBrowser.xaml.cs [hunk 1]
```diff
@@ -262,34 +259,6 @@ private void AllView_PreparingCellForEdit(object sender, DataGridPreparingCellFo
                 return;
             }
 
-            // Only cancel if this event was triggered by a tap
-            // Do not cancel when user presses F2 or context menu
-            if (e.EditingEventArgs is TappedRoutedEventArgs)
-            {
-                if (AppSettings.OpenItemsWithOneclick)
```
Comment (gave92) [inline Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: This is still neded I think, try testing with `OpenItemsWithOneclick` option enabled

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: Sorry

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: @gave92 - before I fix this, I never worked single-click mode. Is there an option at all to rename based on mouse interaction?

Comment (gave92) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:234]: In single-click mode rename is only with "F2" or mouse right click (context menu) <- confirm @yair100?

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: @gave92 - in this case I am not sure why we need this logic in cell editing as it should be already handled in ItemClick but I guess it was added for reason.

Comment (gave92) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: The DataGrid would put the cell in rename mode, can you point me to the ItemClick you are referring to? It should be only in GridView not Listview

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: I was actually referring to `AllView_ItemPress`. 
After few minutes of playing I was able to get to PrepairingEdit in one-click mode.. but it might originate from some other defect...

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: @gave92 - I think there are several issues here:
1. `AllView_ItemPress` should mark event as handled if it decides to invoke ItemClick, so it won't get routed to the Grid  (to avoid editing from being invoked).
2. There RectangleSelection creates another issue - it unregisters `All_ViewItemPress` and.. it is too sensitive and is started even if from my perspective I didn't move the mouse after clicking.  So sometimes when I click, the rectangle selection kicks in, unregisters AllView_ItemPressed and nothing happens upon click.

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: I can return the protection, to edit - but it should never be invoked in normal conditions.

Comment (gave92) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: > I think there are several issues here

Mouse/Pointer/Keyboard event handling is a mess, yeah
1- Ok, if it prevents AllView_PreparingCellForEdit from firing in case of single click mode, then we can remove the whole code section as you did
2- Selection rectangle's SelectionStarted should be only called if the mouse was moved at least a bit

Comment (yaichenbaum) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: > In single-click mode rename is only with "F2" or mouse right click (context menu) <- confirm @yaichenbaum?

That's correct.

Comment (Don-Vito) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: @gave92 - I returned the protection and I am changing the sensitivity of selection in a separate PR

Comment (duke7553) [reply Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:255]: @Don-Vito @gave92 Looking at the [docs](https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/datagrid_guidance/editing_inputvalidation#editing-methods-and-events), it is actually unnecessary to handle the `PreparingCellForEdit` event. We should evaluate whether handling the `BeginningEdit` event will suffice for our special processing since the `BeginningEdit` event does occur **before** a cell or row enters editing mode instead of when a cell enters the editing mode.

Basically since `PreparingCellForEdit` is technically a method as per the docs, it may be bad practice to override it with our own logic.

**Review by duke7553 — COMMENTED** [review]
@Don-Vito @gave92 Looking at the docs, it is actually unnecessary to handle the PreparingCellForEdit event. We should evaluate whether handling the BeginningEdit event will suffice for our special processing since the BeginningEdit event does occur before a cell or row enters editing mode instead of when a cell enters the editing mode.

Basically since PreparingCellForEdit is technically a method as per the docs, it may be bad practice to override it with our own logic.

**Review by duke7553 — DISMISSED** [review]

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito @gave92 Looking at the docs, it is actually unnecessary to handle the PreparingCellForEdit event. We should evaluate whether handling the BeginningEdit event will suffice for our special processing since the BeginningEdit event does occur before a cell or row enters editing mode instead of when a cell enters the editing mode.
> 
> Basically since PreparingCellForEdit is technically a method as per the docs, it may be bad practice to override it with our own logic.

Sounds really good - I can try to move it! Though, probably, it is better to do it in a separate PR, once this one gets merged - to preserve the scope of this one.

**Comment by gave92** [comment]
@Don-Vito sry I was wrong the debounce is still needed. To reproduce:
- select an item
- click outside datagrid
- double click on the file name -> instead of opening it will start renaming

**Comment by Don-Vito** [comment]
@gave92 - the double click rename reproduced to me randomly. But after rebasing and moving the logic to BeginningEdit, it doesn't happen to me anymore.. Can you please recheck?

**Comment by gave92** [comment]
@Don-Vito looks like that issue is resolved 👍 I have a new one though: when single click mode is enabled, sometimes it doesn't work and opening an item requires two clicks, is this something you can reproduce?

**Comment by Don-Vito** [comment]
@gave92 - I will try to repro. But probably we should merge the "calm down rs" first, as it sometimes "steals" clicks by unregistering itempressed

**Comment by gave92** [comment]
> I really hate this solution

Apparently there is an alternative solution to the rename issue. The following code will reset the current datagrid cell.
```cs
uiElement.GetType().GetProperty("CurrentColumnIndex", BindingFlags.NonPublic | BindingFlags.Instance).SetValue(uiElement, -1);
```
So if you added it [here](https://github.com/files-community/Files/blob/master/Files/UserControls/Selection/RectangleSelection_DataGrid.cs#L173) it should work without the need of RootGrid_Tapped/isAllViewCurrentSlotValid/debounce timer

Not sure if you like this better :)

**Comment by Don-Vito** [comment]
> > I really hate this solution
> 
> Apparently there is an alternative solution to the rename issue. The following code will reset the current datagrid cell.
> ```cs
> typeof(DataGrid).GetProperty("CurrentColumnIndex", BindingFlags.NonPublic | BindingFlags.Instance).SetValue(uiElement, -1);
> ```
> So if you added it [here](https://github.com/files-community/Files/blob/5d93f493869902f3e6e0e3e5ada9acac6412ed4c/Files/UserControls/Selection/RectangleSelection_DataGrid.cs#L184) it should work without the need of RootGrid_Tapped/isAllViewCurrentSlotValid/debounce timer
> 
> Not sure if you like this better :)

Cool! This should be much better. I will test it later on. Wondering if we can do it with CurrentItem/Slot.

**Comment by Don-Vito** [comment]
Closing this one in favor of https://github.com/files-community/Files/pull/2813
